// @green-loop/backend/src/routes/checkins.ts
import { Router } from 'express';
import { authMiddleware } from '../middleware/auth.js';
import { sendSuccess, sendError } from '../lib/response.js';
import { validateCoordinates, findNearby } from '../services/spatial.js';
import { isDuplicateCheckIn, awardCheckIn, upsertResident } from '../services/points.js';
import pool from '../db/pool.js';

const router = Router();

router.post('/', authMiddleware, async (req, res, next) => {
  try {
    const { collectionPointId, lat, lng } = req.body as {
      collectionPointId: unknown;
      lat: unknown;
      lng: unknown;
    };

    // Validate coordinates
    const coordCheck = validateCoordinates(lat, lng);
    if (!coordCheck.valid) {
      sendError(res, 400, `Invalid coordinate: ${coordCheck.field}`, coordCheck.field, req.requestId);
      return;
    }

    if (typeof collectionPointId !== 'number') {
      sendError(res, 400, 'collectionPointId must be a number', 'collectionPointId', req.requestId);
      return;
    }

    const resident = req.resident!;

    // Ensure resident exists in DB
    await upsertResident(resident.residentId, resident.displayName, resident.district);

    // Duplicate check-in
    const duplicate = await isDuplicateCheckIn(resident.residentId, collectionPointId);
    if (duplicate) {
      res.status(409).json({
        error: {
          message: 'Already checked in recently',
          retryAfterSeconds: 3600,
          requestId: req.requestId,
        },
      });
      return;
    }

    // Proximity check: find nearby within 200 m
    const nearby = await findNearby(lat as number, lng as number, 200);
    const found = nearby.find((p) => p.id === collectionPointId);
    if (!found) {
      sendError(res, 422, 'Collection point is not within 200 metres of your location', undefined, req.requestId);
      return;
    }

    // Get collection point tier and underserved status from DB
    const cpResult = await pool.query<{ access_tier: 'basic' | 'premium'; density: number | null }>(
      `SELECT
         cp.access_tier,
         COUNT(cp2.id)::double precision / NULLIF(d.area_km2, 0) AS density
       FROM collection_points cp
       JOIN districts d ON d.id = cp.district_id
       LEFT JOIN collection_points cp2 ON cp2.district_id = d.id
       WHERE cp.id = $1
       GROUP BY cp.access_tier, d.area_km2`,
      [collectionPointId]
    );
    if (cpResult.rows.length === 0) {
      sendError(res, 422, 'Collection point not found', undefined, req.requestId);
      return;
    }
    const tier = cpResult.rows[0]!.access_tier;
    const density = cpResult.rows[0]!.density;
    const isUnderserved = density === null || density < 0.5;

    // Award points
    const result = await awardCheckIn(
      resident.residentId,
      collectionPointId,
      tier,
      lat as number,
      lng as number,
      isUnderserved
    );

    sendSuccess(res, {
      pointsAwarded: result.pointsAwarded,
      buildingPointsAwarded: result.buildingPointsAwarded,
      totalPoints: result.totalPoints,
      checkinId: result.checkinId,
    }, 201);
  } catch (err) {
    next(err);
  }
});

export default router;
